### PR TITLE
Clblast fixes + enhancements to save VRAM and offload more layers

### PR DESCRIPTION
**Pull request by LostRuins**
This PR builds on top of occam's changes, https://github.com/ggerganov/llama.cpp/pull/1653, which should be merged first.

Summary: This solves a problem where allocating and freeing gradually increasing sized device buffers in sequence ends up creating a lot of useless buffers that do not get fully utilized, squandering precious VRAM. For example, allocating and then freeing a `ggml_cl_pool_malloc` in the sequence 1gb->2gb->3gb->4gb ends up using a total of 10gb, whereas after this PR only 4gb would be utilized.

Clblast fixes + enhancements to save VRAM and offload more layers:

1. Change all Clblast buffers to `CL_MEM_READ_WRITE`, as the pool malloc currently doesn't properly handle them, since the original flags are not currently preserved.
2. When recycling buffers in pool malloc, always assign the SMALLEST available buffer that fits, instead of the FIRST available buffer.
3. When failing to recycle a buffer in pool malloc (if all existing free buffers are too small), instead recycle the largest available free buffer by resizing it.

With these enhancements I was able to load 2 additional GPU layers into VRAM, at no/minimal observed cost to performance.

This PR is built based off the `master` branch of llama.cpp and follows it's MIT license in full.

### CODE REVIEW

**Comment by YellowRoseCx** [comment]
+1

File: ggml-opencl.cpp [hunk 1]
```diff
@@ -608,15 +609,36 @@ static cl_mem ggml_cl_pool_malloc(size_t size, size_t * actual_size, cl_mem_flag
     scoped_spin_lock lock(g_cl_pool_lock);
     cl_int err;
 
+    int best_i = -1, best_size = std::numeric_limits<size_t>::max(); //smallest unused buffer that fits our needs
```
[the remark this answers is not in the document]
Comment (0cc4m) [reply ggml-opencl.cpp:612]: I think this change is a good idea.

File: ggml-opencl.cpp [hunk 2]
```diff
@@ -577,15 +608,36 @@
     scoped_spin_lock lock(g_cl_pool_lock);
     cl_int err;
 
+    int best_i = -1, best_size = (size_t)-1; //smallest unused buffer that fits our needs
+    int worst_i = -1, worst_size = 0; //largest unused buffer seen so far
```
[the remark this answers is not in the document]
Comment (0cc4m) [reply ggml-opencl.cpp:613]: This is less cluttered than the last one, but also looks better in two lines.

File: ggml-opencl.cpp [hunk 3]
```diff
@@ -692,9 +714,10 @@ static void ggml_cl_mul_f32(const ggml_tensor * src0, const ggml_tensor * src1,
     size_t x_size;
     size_t d_size;
 
-    cl_mem d_X = ggml_cl_pool_malloc(ne0 * sizeof(float), &x_size, CL_MEM_READ_ONLY); // src0
+    cl_mem d_X = ggml_cl_pool_malloc(ne0 * sizeof(float), &x_size, CL_MEM_READ_WRITE); // src0
```
Comment (0cc4m) [inline ggml-opencl.cpp:717]: You could just remove the flags parameter of the malloc function in this case, we don't use it anymore.

**Comment by LostRuins** [comment]
Fixes applied

File: ggml-opencl.cpp [hunk 4]
```diff
@@ -604,21 +605,44 @@ struct cl_buffer {
...
+        *actual_size = b.size;
+        b.size = 0;
+        return mem;
+    }
+    if(worst_i!=-1) //no buffer that fits our needs, resize largest one to save memory
+    {
+         cl_buffer& b = g_cl_buffer_pool[worst_i];
+         cl_mem mem = b.mem;
+         b.size = 0;
+         clReleaseMemObject(mem);
```
Comment (0cc4m) [inline ggml-opencl.cpp:642]:
```suggestion
         clReleaseMemObject(mem);
```

**Review by 0cc4m — CHANGES_REQUESTED** [review]

**Review by 0cc4m — APPROVED** [review]

**Pull request merged**